### ISSUE TITLE
[469b] Trigger with huge CollisionRadius below level produce high CPU load

### ISSUE THREAD

**Issue by SeriousBuggie** [OldUnreal/UnrealTournamentPatches#371]
Reproduce:
1. Start UT on map [MH-TheFifthVortexBeta17_bug2.zip](https://github.com/OldUnreal/UnrealTournamentPatches/files/6412400/MH-TheFifthVortexBeta17_bug2.zip)
2. Open Process Explorer CPU graph for UT process.
3. Wait some time.
4. Run in console `KillAll MyTrigger`

Expected result: CPU load still same.
Actual result: CPU load noticeable drop.
On 12 core CPU drop from 6% to 5%:
![img](https://user-images.githubusercontent.com/70026933/116830537-9effec00-abb3-11eb-8005-403bbca73947.png)

MyTriigger on map only one with CollisionRadius=40000, placed near bottom of all level, and collide with nobody:
![img](https://user-images.githubusercontent.com/70026933/116830584-f7cf8480-abb3-11eb-93ba-a4f3aaa197d8.png)

**Comment by an-eternity** [comment]
Yes, this is very hard problem in all UT99 versions...
And the actual problem is not just a static CPU load when there is an actor with huge collision size in the Level, but performance of the collision-related native functions that appears to be significantly reduced in this case.

Previous v469 versions used to have a collision hash performance problems that resulted in a high performance losses in such cases.
For this reason we had no choice but to reduce the limit for CollisionRadius for all actors to 10000, sizes over 15000 caused serious performance problems. With at least one actor having CollisionRadius of 40000 the server, in some cases, was unable to startup at all.
There were few messages about this problem and the test results on the forum: https://www.oldunreal.com/phpBB3/viewtopic.php?f=42&t=4519&p=43211#p43211
In v469b this performance problem is fixed, the collision hash now is working even faster than in v436..v451. As far as i understand it is already the best what can ever be achieved, and further improvements are barely possible.

So, at this point, it is rather important suggestion for mappers: try to keep the CollisionRadius value below 8000-10000, as exceeding this range may cost of some significant amount of CPU time... In the matter of performance, it is better to add several actors with smaller collision size, instead of one actor with huge collision size.

**Comment by SeriousBuggie** [comment]
Perhaps I don't understand something, but why is it simple not to take into account the minimum and maximum collision height?

**Comment by an-eternity** [comment]
The test cases with 'Trigger' having CollisionRadius in a range 20000..40000 and CollisionHeight in a range 1024..2048 have shown that reducing a CollisionHeight is only effective down to ~64..128, reducing it further even down to 2 units gives almost no performance improvement, and the performance loss produced by the huge CollisionRadius was still very high.

---

In the past year, when i reported about strange performance losses on the map MH-TheFifthVortexBeta, it was (as we later discovered) caused exactly by the combination of these 2 problems: the problem of huge CollisionRadius (there are several of such huge Triggers in the map), and the collision hash performance problem in older v469 builds.

Thanks to fixed/improved collision hash performance in v469b these losses are not that high as before, and the map MH-TheFifthVortexBeta is working fine now (even if the fix in GameType that limits collision sizes is disabled).

**Comment by SeriousBuggie** [comment]
In this example `MyTrigger` has `CollisionHeight` = 40.
And I still not understand why UT not use some fast check of min and max Z coordinate of collision for exclude such triggers from heavy calculation.

**Comment by an-eternity** [comment]
Yes, it would be nice to have some optional fast check mode for such triggers that only perform a very simple tasks without any high precision requirements, if this kind of enhancement may even be practically possible... So that performance losses produced by these huge collision size triggers would be reduced even more, down to insignificant levels.

**Comment by SeriousBuggie** [comment]
I try figure what produce CPU load.
Delete all trigger except MyTrigger, start level:
```
> open autoplay
> killall mytrigger
> open autoplay
> killall brush
> killall scriptedpawn
> killall flockpawn
> killall decal
> killall decoration
> killall inventory
> killall keypoint
> killall light
> killall navigationpoint
> killall mytrigger
```
Even if most actors killed run last line noticeable drop CPU usage.
So high CPU usage connected with MyTrigger self, without involve another actors which can be deleted at runtime.
Or actors already produce huge structures, which need CPU for iterate over it.

---
In UnrealEd select all Brushes+ via `actor select ofsubclass class=brush` and delete them.
Save, Run, `killall mytrigger` - no any CPU drop.

---
In UnrealEd select all Brushes via `actor select ofclass class=brush` and delete them.
Save, Run, `killall mytrigger` -  CPU drop.

---
In UnrealEd select all Movers via `actor select ofsubclass class=mover` and delete them.
Save, Run, `killall mytrigger` - no any CPU drop.

Interested thing - `killall Mover` not reduce CPU load.

**Comment by SeriousBuggie** [comment]
If i Delete Mover158 no CPU drop when delete MyTrigger.
It is big Mover at bottom level with bCollideActors.
![img](https://user-images.githubusercontent.com/70026933/116911969-3c106280-ac50-11eb-8c00-1b2bddc4fe34.png)
![img](https://user-images.githubusercontent.com/70026933/116911986-4599ca80-ac50-11eb-8cb0-2bfe564a2365.png)

Without bCollideActors CPU load on start level noticeable less. But anyway on kill MyTrigger CPU drop.

Move this Mover in middle of level not reduce CPU load on start, and not remove CPU drop when kill MyTrigger.

**Comment by SeriousBuggie** [comment]
If I stretch this mover on almost all level by XY coords, then CPU load on start level highly increased and highly dropped when I delete MyTrigger.
If I reduce XY size then CPU on load low and not drop when I delete MyTrigger.

So my conclusion: big collision radius for actor + big mover = high CPU load.
Possible this info can be used for eliminate bottleneck in the engine.

**Comment by an-eternity** [comment]
In the past year when i inspected this problem with the map MH-TheFifthVortexBeta and with older v469 build, i noticed that CPU load gets significantly increased exactly at the moment when those flying Movers (in the "sky land" location) begin an infinite movement cycle up/down... When i stopped them and they were no longer moving, CPU load remained increased. Even when i deleted those movers, it didn't make any effect on the CPU load - it remained increased. What exactly was causing this effect, - those movers and the fact that they begin to move, or maybe something else that happens in the map in that moment, - is still question i have no answer for...

**Comment by an-eternity** [comment]
Interesting test case with big Mover provided by this map: https://unrealarchive.org/maps/unreal-tournament/monster-hunt/M/mh-monstermash2_7187cd7f.html
If, for example, you will trigger an event 'plat2', and after a short moment trigger an event 'blok' (it will try to spawn Projectiles inside of the Mover), then CPU load will raise to maximum for several seconds. This may be useful for some specific test purposes, such unusual cases of collision with Mover normally don't happen in most of the maps.

**Comment by stijn-volckaert** [comment]
This is a fundamental problem in UE1. UE1 uses a collision hash for collision detection. The level is divided into a big 3D grid and every box/bucket in the grid contains a list of pointers to actors that overlap with that box. This means that the number of boxes that link to a specific actor is directly proportional to the size of that actor. It also means that collision testing/tracing is extremely expensive if the grid contains large actors. It becomes even more expensive if the grid contains large actors _and_ the actor you're trying to move is also large. We could solve this by backporting the octree collision code from Unreal 227.

**Comment by an-eternity** [comment]
> We could solve this by backporting the octree collision code from Unreal 227.

This would be nice. We are looking forward for this. Hope to have this already in v469c or in v469d.

**Comment by CacoFFF** [comment]
The collision hash isn't really designed to handle oversized objects.
I temporarily solved that (and other collision issues) in XC_Engine using a hand tweaked grid, but said code isn't friendly to non-X86 platforms, not something i'd port into the main game.

**Comment by SeriousBuggie** [comment]
Need look into wall and ensure your FPS is reach fps limit. If it is not - need correct fps limit.
Or else you can't see difference after drop trigger.
